learn/shorts/functional-programming-python.qmd

```diff
@@ -0,0 +1,228 @@
+when working with data transformations.
+
+In this lesson, we will cover three built-in functions in Python that
+are commonly used in functional programming: `filter`, `map`, and
+`reduce`. We will also discuss the `lambda` keyword, which allows for
+the creation of small anonymous functions. These functions are often
+used in conjunction with `filter`, `map`, and `reduce` to create more
+concise and readable code.
+
+### **`filter`**
+
+This is a built-in function in Python that takes a function and an
+iterable as arguments and returns an iterator over the elements of the
+iterable for which the function returns `True`.
+
+``` python
+filter(function, iterable)
+```
+
+The `function` should return `True` or `False` for each element in
+`iterable`. Only elements for which `function` returns `True` are
+included in the output.
+
+For example:
+
+``` python
+numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+even_numbers = list(filter(lambda x: x % 2 == 0, numbers))
+print(even_numbers)
+```
+
+In the example above, the `filter` function is used to filter out the
+even numbers from the list of numbers. The lambda function is used to
+define the condition for filtering, in this case, checking if the number
+produces a remainder of 0 when divided by 2. The `filter` function on
+its own returns what is called an iterator, so in order for us to see
+the result, we need to convert it to a list. We do the same below with
+the `map` function.
+
+### **`map`**
+
+This is a built-in function in Python that takes a function and an
+iterable as arguments and returns an iterator over the results of
+applying the function to the elements of the iterable.
+
+``` python
+map(function, iterable)
+```
+
+The `function` is applied to each element of `iterable`, and the results
+are returned as an iterator.
+
+For example:
+
+``` python
+def square(x):
+    return x ** 2
+
+numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+squared_numbers = list(map(square, numbers))
+print(squared_numbers)
+```
+
+In this example, the `map` function is used to square each number in the
+list of numbers. The lambda function is used to define the operation to
+be applied to each element of the list, in this case by squaring it. The
+`map` function returns an iterator, so we convert it to a list to see
+the result.
+
+#### **`reduce`**
+
+This is a function in the `functools` module in Python that takes a
+function and an iterable as arguments and returns a single value by
+applying the function cumulatively to the elements of the iterable.
+
+``` python
+reduce(function, iterable)
+```
+
+The `function` takes two arguments and is applied cumulatively to the
+elements of `iterable` to reduce them to a single value.
+
+It is slightly different from the first two functions in that it isn't
+built into Python but is available in the `functools` module.
+
+For example:
+
+``` python
+from functools import reduce
+numbers = [1, 2, 3, 4, 5]
+sum_of_numbers = reduce(lambda x, y: x + y, numbers)
```

Comment:
This feels like an inconsistent behaviour that the others need `list()` and this one doesn't. (just a random comment)